pyquil/device.py

```diff
@@ -319,9 +323,56 @@ def from_dict(d):
         )
 
 
-class Device(object):
+def isa_from_graph(graph: nx.Graph, oneq_type='Xhalves', twoq_type='CZ') -> ISA:
+    """
+    Generate an ISA object from a NetworkX graph.
+
+    :param graph: The graph
+    :param oneq_type: The type of 1-qubit gate. Currently 'Xhalves'
+    :param twoq_type: The type of 2-qubit gate. One of 'CZ' or 'CPHASE'.
+    """
+    all_qubits = list(range(max(graph.nodes) + 1))
+    qubits = [Qubit(i, type=oneq_type, dead=i not in graph.nodes) for i in all_qubits]
+    edges = [Edge([a, b], type=twoq_type, dead=False) for a, b in graph.edges]
+    return ISA(qubits, edges)
+
+
+def isa_to_graph(isa: ISA) -> nx.Graph:
+    """
+    Construct a NetworkX qubit topology from an ISA object.
+
+    This discards information about supported gates.
+
+    :param isa: The ISA.
+    """
+    return nx.from_edgelist(e.targets for e in isa.edges if not e.dead)
+
+
+class AbstractDevice(ABC):
+
+    def qubit_topology(self) -> nx.Graph:
```

Comment:
@abstractmethod instead of throwing NotImplemented

Reply:
0d3d5e132dee1a2318eac31c5cb88a387e9a60e6